### PR TITLE
Prow Plugin to Block Pull Requests with Merge Commits

### PR DESCRIPTION
**Pull request by alisondy**
Adding a plugin to prow that will block Pull Requests with Merge Commits
A Merge Commit is defined as a commit with more than one Parent commit
Resolves : #5376

### CODE REVIEW

File: prow/plugins/mergecommitblock/mergecommitblock.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package mergecommitblock adds a do-not-merge label to pull requests which contain merge commits
+// Merge commits are defined as commits that contain more than one parent commit SHA
+
+package mergecommitblock
+
+const pluginName = "block-mc"
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblock/mergecommitblock.go]: This is going to be user-facing, so I think we should prefer `mergecommitblocker` or similar.

**Comment by alisondy** [comment]
apologies, will pick this up again, unless someone else would like to.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	if err != nil {
+
+	}
+	// Iterate through them and check for parent commits
+	var needsLabel bool = false
+
+	for _, commit := range commits {
+		if len(commit.Parents) > 1 {
+			needsLabel = true
+			continue
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: did you mean `break`?

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		}
+	}
+
+	// Once finished iterating, Label if merge commits were identified
+	issueLabels, err := gc.GetIssueLabels(owner, repo, num)
+	if err != nil {
+		le.Warnf("while retrieving labels, error: %v", err)
+	}
+
+	f := func(label string, labels []github.Label) bool {
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: I'm not sure if this closure is necessary

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+// isPRChanged takes a github Pull request event and returns a boolean value, which indicates if code diffs have changed
+func isPRChanged(pe github.PullRequestEvent) bool {
+	switch pe.Action {
+	case github.PullRequestActionOpened:
+		return true
+	case github.PullRequestActionReopened:
+		return true
+	case github.PullRequestActionSynchronize:
+		return true
+	case github.PullRequestActionEdited:
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: Is this a PR body edit? Wouldn't we get a synchronization event on code push?

File: prow/plugins/mergecommitblocker/mergecommitblocker_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mergecommitblocker
+
+import (
+	"fmt"
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker_test.go]: nit: we try to group our imports, `stdlib`, third-party, then internal

```go
import (
	"fmt"
	"reflect"
	"testing"

	"github.com/sirupsen/logrus"

	"k8s.io/test-infra/prow/github"
	"k8s.io/test-infra/prow/github/fakegithub"
)
```

File: prow/plugins/mergecommitblocker/mergecommitblocker_test.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/github/fakegithub"
+	"reflect"
+	"testing"
+)
+const (
+	helpWanted = "help-wanted"
+	mergeCommit = "do-not-merge/contains-merge-commits"
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker_test.go]: Why not import this?

File: prow/plugins/mergecommitblocker/mergecommitblocker_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	var testcases = []struct {
+		name string
+		pullRequestEvent github.PullRequestEvent
+		commits          []github.RepositoryCommit
+		initialLabels  []string
+		addedLabel string
+		removedLabel   string
+	}{
+		{
+			name: "should label with do-not-merge/contains-merge-commits when merge commits present",
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker_test.go]: I think you're missing cases for no-ops when the label is needed and exists, or is not needed and doesn't exist, or when GitHub gives errors. But these tests are sweet!

File: prow/plugins/mergecommitblocker/mergecommitblocker.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	}
+	// Store all info about the owner, repo, and num of pull request
+	var (
+		owner = pe.PullRequest.Base.Repo.Owner.Login
+		repo  = pe.PullRequest.Base.Repo.Name
+		num   = pe.PullRequest.Number
+	)
+
+	// Use github client to get the commits in the pull request
+	commits, err := gc.ListPRCommits(owner, repo, num)
```
Comment (stevekuznetsov) [inline prow/plugins/mergecommitblocker/mergecommitblocker.go]: I'm worried this will eat up tokens in a major way -- can we use the `git` client instead?

Comment (alisondy) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: Should be possible,
Ran this git command gave me commit hashes, when there were merge commits two hashes were present
```git --no-pager log master...$(git rev-parse --abbrev-ref HEAD) --pretty=%P``
Should be able to use ```git``` client to replicate this ✨

Comment (stevekuznetsov) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: You should be able to use

```
git log --min-parents=2 $( git merge-base <pr-merged-into-branch-ref> <target-branch-ref> )..HEAD
```

Or similar

File: prow/plugins/mergecommitblocker/mergecommitblocker_test.go (hunk 6 above)
Comment (alisondy) [reply prow/plugins/mergecommitblocker/mergecommitblocker_test.go]: i will admit i was copying part of another plugins tests 😎will fix

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 8 above)
Comment (alisondy) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: hello there not dead, so I would alter the git client to do a git log? Thanks in advance 🙏

File: prow/plugins/mergecommitblocker/mergecommitblocker_test.go (hunk 5 above)
Comment (alisondy) [reply prow/plugins/mergecommitblocker/mergecommitblocker_test.go]: ✨

File: prow/labels/labels.go [hunk 9]
```diff
@@ -34,6 +34,7 @@ const (
 	LifecycleFrozen = "lifecycle/frozen"
 	LifecycleRotten = "lifecycle/rotten"
 	LifecycleStale  = "lifecycle/stale"
+	MergeCommits	= "do-not-merge/contains-merge-commits"
```
Comment (stevekuznetsov) [inline prow/labels/labels.go]: I think you need to run `gofmt`?

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 4 above)
Comment (stevekuznetsov) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: This is outstanding -- I don't think we need to check the commits on an edit event, that's just someone changing their PR title or body.

File: prow/plugins/mergecommitblocker/mergecommitblocker.go (hunk 8 above)
Comment (stevekuznetsov) [reply prow/plugins/mergecommitblocker/mergecommitblocker.go]: The `git` client right now takes actions but doesn't try to create `commit` objects or anything, so we could keep the amount of code necessary small by creating a method like:

```go
func (r *Repo) MergeCommitsExistBetween(target, head string) (bool, error) {
```

Where that function would do the equivalent of the command I pasted above and just returns whether or not a merge commit was found.

File: prow/plugins/mergecommitblocker/mergecommitblocker_test.go (hunk 7 above)
Comment (stevekuznetsov) [reply prow/plugins/mergecommitblocker/mergecommitblocker_test.go]: I think we're still outstanding on those other test cases

**Comment by alisondy** [comment]
has been fixed

**Pull request closed without merging**